app/models/stage.rb

```diff
@@ -252,19 +243,25 @@ def validate_not_auto_deploying_without_buddy
     end
   end
 
-  # has to be done after command_ids assignment is done
-  def append_new_command
-    return if @command.blank?
-    new_command = project.commands.new(command: @command)
-    previous = stage_commands.map(&:position).max || 0
-    stage_commands.build(command: new_command, position: previous + 1)
-  end
-
   def environment_lock?(lock)
     lock.resource_type == "Environment" && environments.any? { |e| lock.resource_equal?(e) }
   end
 
   def project_lock?(lock)
     lock.resource_type == "Project" && lock.resource_equal?(project)
   end
+
+  def find_or_create_stage_command(command_or_id, position)
+    command_or_id = command_or_id.to_s
+
+    if existing_command = command_or_id.match?(/\A\d+\z/) && Command.find_by(id: Integer(command_or_id))
```

Comment:
n+1 ... only rely on `.detect`

Reply:
if you want to make sure it exists, could do `ids = Command.pluck(:id)` ... or some other way of making sure typing an integer does not mess it up :D